### PR TITLE
[Remove SitesList] Remove Site#isUpgradeable()

### PR DESCRIPTION
**Pull request by Tug**
This PR goal is to remove `Site#isUpgradeable()`.

This will help toward making the `Site` instance almost identical to the site objects in the redux store so the transition from `SitesList` to `state/sites` is easier. Progressively removing methods from `Site` and `SitesList` is a good way to help with this transition.

### Testing Instructions
- Boot this branch locally (or using calypso.live)
- Go to http://calypso.localhost:3000/sites and assert that there is no error when using this page
- Go to http://calypso.localhost:3000/domains assert that there is no error when using this page
- Go to `http://calypso.localhost:3000/domains/add/:exisiting-site-slug` and assert that there is no error
- Go to `http://calypso.localhost:3000/domains/add/suggestion/hello+world/:exisiting-site-slug` and assert that there is no error
- Go to `http://calypso.localhost:3000/domains/add/site-redirect/:exisiting-site-slug` and assert that there is no error

### Reviews
- [ ] Code
- [ ] Product

### CODE REVIEW

File: client/my-sites/sites/sites.jsx [hunk 1]
```diff
@@ -12,8 +14,9 @@ import Main from 'components/main';
 import observe from 'lib/mixins/data-observe';
 import SiteSelector from 'components/site-selector';
 import { addSiteFragment } from 'lib/route';
+import { isSiteUpgradeable } from 'state/sites/selectors';
 
-export default React.createClass( {
+const Sites = React.createClass( {
 	displayName: 'Sites',
```
Comment (tyxla) [inline client/my-sites/sites/sites.jsx]: It appears we can remove the `displayName` now.

File: client/my-sites/upgrades/domain-search/domain-search.jsx [hunk 2]
```diff
@@ -165,9 +166,12 @@ var DomainSearch = React.createClass( {
 
 module.exports = connect(
 	function( state, props ) {
+		const selectedSite = props.sites.getSelectedSite();
```
Comment (tyxla) [inline client/my-sites/upgrades/domain-search/domain-search.jsx]: I'd suggest that we don't introduce another instance of `sitesList.getSelectedSite()` - we're trying to get rid of them, see #8726. How about using the `getSelectedSite()` selector instead:

```
import { getSelectedSite } from 'state/ui/selectors';
...
const selectedSite = getSelectedSite( state );
```

File: client/my-sites/upgrades/domain-search/site-redirect.jsx [hunk 3]
```diff
@@ -2,17 +2,19 @@
  * External dependencies
  */
 var page = require( 'page' ),
-	React = require( 'react' );
+	React = require( 'react' ),
+	{ connect } = require( 'react-redux' );
```
Comment (tyxla) [inline client/my-sites/upgrades/domain-search/site-redirect.jsx]: AFAIK we don't encourage using `require()` just for consistency reasons. Refactoring the entire component to use ES6 imports can definitely be done in another PR, but we should only use `import` for any new imports we introduce.

File: client/my-sites/upgrades/domain-search/site-redirect.jsx [hunk 4]
```diff
@@ -60,4 +62,12 @@ var SiteRedirect = React.createClass( {
 	}
 } );
 
-module.exports = SiteRedirect;
+export default connect(
+	( state, ownProps ) => {
+		const selectedSite = ownProps.sites.getSelectedSite();
```
Comment (tyxla) [inline client/my-sites/upgrades/domain-search/site-redirect.jsx]: Just like my previous comment above, I'd suggest that we don't introduce another instance of `sitesList.getSelectedSite()`. Then we'll also no longer need to pass `ownProps` to `mapStateToProps`.

File: client/my-sites/upgrades/map-domain/index.jsx [hunk 5]
```diff
@@ -1,26 +1,29 @@
 /**
  * External dependencies
  */
-var page = require( 'page' ),
-	React = require( 'react' ),
-	omit = require( 'lodash/omit' ),
-	{ connect } = require( 'react-redux' );
+import page from 'page';
+import React from 'react';
+import omit from 'lodash/omit';
```
Comment (tyxla) [inline client/my-sites/upgrades/map-domain/index.jsx]: Might be a good idea to also convert such instances to use root Lodash imports, for example:

```
import { omit } from 'lodash';
```

See #6539 for more information.

File: client/my-sites/upgrades/domain-search/site-redirect.jsx [hunk 6]
```diff
@@ -35,7 +37,7 @@ var SiteRedirect = React.createClass( {
 	checkSiteIsUpgradeable: function( ) {
 		var selectedSite = this.props.sites.getSelectedSite();
```
Comment (tyxla) [inline client/my-sites/upgrades/domain-search/site-redirect.jsx]: This looks like the perfect place to migrate this to use the `getSelectedSite` selector, once you change the usage in `mapStateToProps` to use it instead of `sitesList`.

File: client/my-sites/upgrades/map-domain/index.jsx [hunk 7]
```diff
@@ -141,8 +144,13 @@ var MapDomain = React.createClass( {
 	}
 } );
 
-module.exports = connect( state => (
-	{
-		domainsWithPlansOnly: currentUserHasFlag( state, DOMAINS_WITH_PLANS_ONLY )
+export default connect(
+	( state, ownProps ) => {
+		const selectedSite = ownProps.sites.getSelectedSite();
```
Comment (tyxla) [inline client/my-sites/upgrades/map-domain/index.jsx]: Just like my previous comment above, I'd suggest that we don't introduce another instance of `sitesList.getSelectedSite()`. Then we'll also no longer need to pass `ownProps` to `mapStateToProps`.

File: client/my-sites/upgrades/map-domain/index.jsx [hunk 8]
```diff
@@ -58,7 +61,7 @@ var MapDomain = React.createClass( {
 
 		const selectedSite = this.props.sites.getSelectedSite();
```
Comment (tyxla) [inline client/my-sites/upgrades/map-domain/index.jsx]: This also looks like the perfect place to migrate this to use the `getSelectedSite` selector.

File: client/state/sites/selectors.js [hunk 9]
```diff
@@ -281,6 +282,24 @@ export function isSitePreviewable( state, siteId ) {
...
 /**
+ * Returns true if the site can be upgraded by the user, false if the
+ * site cannot be upgraded, or null if upgrade ability cannot be
+ * determined.
+ *
+ * @param  {Object}   state  Global state tree
+ * @param  {Number}   siteId Site ID
+ * @return {?Boolean}        Whether site is upgradeable
+ */
+export function isSiteUpgradeable( state, siteId ) {
```
Comment (tyxla) [inline client/state/sites/selectors.js]: We should also consider adding tests for this selector.

File: client/state/sites/selectors.js [hunk 10]
```diff
@@ -281,6 +282,24 @@ export function isSitePreviewable( state, siteId ) {
...
+	// Cannot determine site upgradeability if there is no current user
+	if ( ! state.currentUser || ! getRawSite( state, siteId ) ) {
```
Comment (tyxla) [inline client/state/sites/selectors.js]: I think we should not be directly accessing another state subtree here - how about using one of the `getCurrentUser` / `getCurrentUserId` selectors instead?

File: client/my-sites/upgrades/domain-search/domain-search.jsx (hunk 2 above)
Comment (Tug) [inline client/my-sites/upgrades/domain-search/domain-search.jsx]: I didn't really introduced it, simply shared it with `sitePlans` below. Using `getSelectedSite` here would be better indeed, but it requires more change like removing the `sites` from props and then removing mixins... I'll try to do it for this file, it will indeed look better 👍

File: client/my-sites/upgrades/domain-search/site-redirect.jsx [hunk 11]
```diff
@@ -1,63 +1,80 @@
...
+import { getSelectedSite, getSelectedSiteId } from 'state/ui/selectors';
+import QueryProductsList from 'components/data/query-products-list';
 
-var SiteRedirect = React.createClass( {
-	mixins: [ observe( 'productsList', 'sites' ) ],
-
-	propTypes: {
+class SiteRedirect extends Component {
+	static propTypes = {
+		isSiteUpgradeable: React.PropTypes.bool,
```
Comment (yurynix) [inline client/my-sites/upgrades/domain-search/site-redirect.jsx]: selectedSite missing from `propTypes`

File: client/my-sites/upgrades/map-domain/index.jsx [hunk 12]
```diff
@@ -1,72 +1,68 @@
...
 
-var MapDomain = React.createClass( {
-	mixins: [ observe( 'productsList', 'sites' ) ],
+const wpcom = wp.undocumented();
 
-	propTypes: {
+class MapDomain extends Component {
+	static propTypes = {
 		query: React.PropTypes.string,
+		isSiteUpgradeable: React.PropTypes.bool,
```
Comment (yurynix) [inline client/my-sites/upgrades/map-domain/index.jsx]: `selectedSite` missing from propTypes

File: client/state/sites/selectors.js [hunk 13]
```diff
@@ -281,6 +283,24 @@ export function isSitePreviewable( state, siteId ) {
...
+ * site cannot be upgraded, or null if upgrade ability cannot be
+ * determined.
+ *
+ * @param  {Object}   state  Global state tree
+ * @param  {Number}   siteId Site ID
+ * @return {?Boolean}        Whether site is upgradeable
+ */
+export function isSiteUpgradeable( state, siteId ) {
+	// Cannot determine site upgradeability if there is no current user
+	if ( ! getCurrentUserId( state ) || ! getRawSite( state, siteId ) ) {
```
Comment (yurynix) [inline client/state/sites/selectors.js]: For my own understanding, why is `getRawSite` check required?

As I understand, we suppose to check: 
`canCurrentUser(...)` which is `state.currentUser.capabilities[ siteId ][ capability ]`. 
that calls `isValidCapability` which returns `null` if `! state.currentUser.capabilities[ siteId ]`,
so that seems already be covered...

Comment (Tug) [inline client/state/sites/selectors.js]: This is mostly to cover all cases for tests, the problem with this `getCurrentUserId` selector is that it depends on another independent subtree of the state so if the whole state tree is not consistent it might lead to undefined behaviors.

Comment (yurynix) [inline client/state/sites/selectors.js]: Thank you for clarifying!

File: client/my-sites/upgrades/domain-search/domain-search.jsx [hunk 14]
```diff
@@ -158,25 +148,19 @@ var DomainSearch = React.createClass( {
...
-				}
-			}
-		};
-	}
-)( DomainSearch );
+}
+
+export default connect(
+	( state ) => ( {
+		selectedSite: getSelectedSite( state ),
```
Comment (tyxla) [inline client/my-sites/upgrades/domain-search/domain-search.jsx]: Since we're calling `getSelectedSite()` more than once here, perhaps we should be storing it in a `const`?

File: client/my-sites/upgrades/domain-search/domain-search.jsx [hunk 15]
```diff
@@ -1,95 +1,85 @@
...
 			this.removeDomain( suggestion );
 		}
-	},
+	}
 
 	handleAddMapping( suggestion ) {
 		upgradesActions.addItem( cartItems.domainMapping( { domain: suggestion.domain_name } ) );
-		page( '/checkout/' + this.props.sites.getSelectedSite().slug );
-	},
+		page( '/checkout/' + this.props.selectedSite.slug );
```
Comment (tyxla) [inline client/my-sites/upgrades/domain-search/domain-search.jsx]: Does it make sense to alter this to use the `getSelectedSiteSlug` selector for this one, so it'll be slightly more resistant to possible future structural changes to the `site` object?

File: client/my-sites/upgrades/domain-search/site-redirect.jsx [hunk 16]
```diff
@@ -1,63 +1,82 @@
...
 			page.redirect( '/domains/add' );
 		}
-	},
+	}
 
-	backToDomainSearch: function() {
-		page( '/domains/add/' + this.props.sites.getSelectedSite().slug );
-	},
+	handleBackToDomainSearch() {
+		page( '/domains/add/' + this.props.selectedSite.slug );
```
Comment (tyxla) [inline client/my-sites/upgrades/domain-search/site-redirect.jsx]: Does it make sense to alter this to use the `getSelectedSiteSlug` selector for this one, so it'll be slightly more resistant to possible future structural changes to the `site` object?

File: client/my-sites/upgrades/map-domain/index.jsx [hunk 17]
```diff
@@ -76,73 +77,87 @@ var MapDomain = React.createClass( {
...
 				productSlug: suggestion.product_slug,
 				domain: suggestion.domain_name
 			} )
 		);
 
-		if ( this.isMounted() ) {
-			page( '/checkout/' + this.props.sites.getSelectedSite().slug );
-		}
-	},
+		page( '/checkout/' + selectedSite.slug );
```
Comment (tyxla) [inline client/my-sites/upgrades/map-domain/index.jsx]: Does it make sense to alter this to use the `getSelectedSiteSlug` selector for this one, so it'll be slightly more resistant to possible future structural changes to the `site` object?

File: client/my-sites/upgrades/map-domain/index.jsx [hunk 18]
```diff
@@ -76,73 +77,87 @@ var MapDomain = React.createClass( {
...
 		// For VIP sites we handle domain mappings differently
 		// We don't go through the usual checkout process
 		// Instead, we add the mapping directly
 		if ( selectedSite.is_vip ) {
-			wpcom.addVipDomainMapping( selectedSite.ID, domain ).then( () => {
-				page( paths.domainManagementList( selectedSite.slug ) );
-			}, error => this.setState( { errorMessage: error.message } ) );
+			wpcom.addVipDomainMapping( selectedSite.ID, domain )
+				.then(
+					() => page( paths.domainManagementList( selectedSite.slug ) ),
```
Comment (tyxla) [inline client/my-sites/upgrades/map-domain/index.jsx]: Does it make sense to alter this to use the `getSelectedSiteSlug` selector for this one, so it'll be slightly more resistant to possible future structural changes to the `site` object?

File: client/my-sites/upgrades/map-domain/index.jsx [hunk 19]
```diff
@@ -76,73 +77,87 @@ var MapDomain = React.createClass( {
...
 			return;
 		}
 
 		upgradesActions.addItem( cartItems.domainMapping( { domain } ) );
 
-		if ( this.isMounted() ) {
-			page( '/checkout/' + selectedSite.slug );
-		}
-	},
+		page( '/checkout/' + selectedSite.slug );
```
Comment (tyxla) [inline client/my-sites/upgrades/map-domain/index.jsx]: Does it make sense to alter this to use the `getSelectedSiteSlug` selector for this one, so it'll be slightly more resistant to possible future structural changes to the `site` object?

File: client/lib/site/index.js [hunk 20]
```diff
@@ -212,10 +212,6 @@ Site.prototype.fetchUsers = function() {
 	}.bind( this ) );
 };
 
-Site.prototype.isUpgradeable = function() {
-	return this.capabilities && this.capabilities.manage_options;
-};
-
```
Comment (stephanethomas) [inline client/lib/site/index.js]: You may want to remove the mention to this method from the [`README.md` file](https://github.com/Automattic/wp-calypso/blob/d9edb63952c6335f9fa8ceab64061af27b6f9694/client/lib/site/README.md).

File: client/my-sites/sites/sites.jsx [hunk 21]
```diff
@@ -44,7 +45,7 @@ export default React.createClass( {
 
 		// Filter out sites with no upgrades on particular routes
 		if ( /^\/domains/.test( path ) || /^\/plans/.test( this.props.sourcePath ) ) {
-			return site.isUpgradeable();
+			return this.props.isSiteUpgradeable( site.ID );
```
Comment (stephanethomas) [inline client/my-sites/sites/sites.jsx]: This new prop should be specified in the prop types above.

**Comment by Tug** [comment]
@tyxla Thanks for reviewing, the change you suggested were good so I did them 👍 

@stephanethomas Indeed I do have commits that mix `isUpgradeable` changes with other refactors but simple things like converting `require` calls to `import` statements and using `const` instead of `var` at a couple places. The idea behind it is that we shouldn't mix `require` with `import` so when it was necessary I converted those. I didn't except this PR to become that big but since it updated old legacy modules I was asked to update those modules fully to ES6. Also keep in mind that the whole PR is a refactor and that there is no core logic change, only modules that get updated to use the new syntax and use new selectors.

**Comment by Tug** [comment]
This is ready for another review.
Sorry for the large PR!

File: client/my-sites/upgrades/domain-search/domain-search.jsx [hunk 22]
```diff
@@ -1,89 +1,82 @@
...
 			this.removeDomain( suggestion );
 		}
-	},
+	}
 
 	handleAddMapping( domain ) {
 		upgradesActions.addItem( cartItems.domainMapping( { domain } ) );
-		page( '/checkout/' + this.props.sites.getSelectedSite().slug );
-	},
+		page( '/checkout/' + this.props.selectedSite.slug );
```
Comment (tyxla) [inline client/my-sites/upgrades/domain-search/domain-search.jsx]: Avoiding usage of `selectedSite.slug` when connected can prevent us from having errors when the site object hasn't been loaded yet - I'd recommend using the `getSelectedSiteSlug` selector for that one. I can also spot this one several other files that this PR affects.

File: client/my-sites/upgrades/domain-search/domain-search.jsx [hunk 23]
```diff
@@ -149,13 +142,17 @@ var DomainSearch = React.createClass( {
...
+}
 
-module.exports = connect(
+export default connect(
 	( state ) => ( {
-		domainsWithPlansOnly: currentUserHasFlag( state, DOMAINS_WITH_PLANS_ONLY )
+		selectedSite: getSelectedSite( state ),
+		domainsWithPlansOnly: currentUserHasFlag( state, DOMAINS_WITH_PLANS_ONLY ),
+		isSiteUpgradeable: isSiteUpgradeable( state, getSelectedSiteId( state ) ),
+		productsList: state.productsList.items,
```
Comment (tyxla) [inline client/my-sites/upgrades/domain-search/domain-search.jsx]: We should definitely introduce a selector for this one (in a separate PR, of course).

**Comment by Tug** [comment]
@tyxla Thanks for reviewing and for your input!
We can do that. There is however a chain of dependencies in the splitting you suggested so we will have to do it in several steps:
```
                        ES6ify `DomainSearch` -- Migrate `DomainSearch`
                    /                                                   \
`isSiteUpgradeable` --  ES6ify `SiteRedirect` -- Migrate `SiteRedirect` --  Cleanup `SitesList`
                    \                                                   /
                           ES6ify `MapDomain` -- Migrate `MapDomain`
```

So 2 or 3 steps at least (we should be able to group ES6ify and migrate tasks together).
I started with the first task here: https://github.com/Automattic/wp-calypso/pull/12059
Feel free to review it (it will have a description soon).

**Comment by tyxla** [comment]
@Tug thanks for that! I've reviewed #12059, feel free to ping me for review of subsequent PRs.

**Comment by Tug** [comment]
Second run, I created the following PRs that can be reviewed in parallel:
- https://github.com/Automattic/wp-calypso/pull/12941 ( DomainSearch, WIP )
- https://github.com/Automattic/wp-calypso/pull/12942 ( SiteRedirect, WIP )
- https://github.com/Automattic/wp-calypso/pull/12943 ( MapDomain, WIP )
- https://github.com/Automattic/wp-calypso/pull/12944 ( Sites )

**Comment by tyxla** [comment]
Awesome @Tug! Will make sure to review them all this week.

**Comment by alisterscott** [comment]
Closing as this PR is stale - please reopen if necessary
